### ISSUE TITLE
Make backpropagation cost function aware

### ISSUE THREAD

**Issue by vanderhe** [vanderhe/fortnet#27]
**Feature Request:**

The backpropagation algorithm is currently implemented with hardcoded loss function gradients. This limitation inevitably means that the choice of loss function only has an impact on training runs with an optimizer that actually infers information based on the loss functions value, i.e. CG and L-BFGS. Unfortunately, this will result in a large number of regression tests having to be recalculated.
